### PR TITLE
Add three new inference methods

### PR DESCRIPTION
**Pull request by denkle**
Still struggle somewhat with "Iterative Hard Thresholding" convergence but committing the changes for now so would appreciate someones fresh eyes on the logic of the inference

### CODE REVIEW

File: sparsecoding/inference.py [hunk 1]
```diff
@@ -670,3 +670,269 @@ def infer(self, data, dictionary, coeff_0=None):
...
+        K = np.ceil(self.sparsity*n_basis).astype(int)
+        
+        # Get dictionary norms in case atoms are not normalized 
+        dictionary_norms = torch.norm(dictionary, p=2, dim=0,keepdim=True)        
+        
+        # Initialize coefficients for the whole batch
+        coefficients = torch.zeros(
+            batch_size, n_basis, requires_grad=False, device=device)
+        
+        # For each sample in the batch
```
Comment (alvinzz) [inline sparsecoding/inference.py:903]: Vectorized:
```
residual = data.clone()  # [batch_size, n_features]
# The basis functions that are used to infer the coefficients will be updated each iteration.
used_basis_fns = torch.zeros((batch_size, n_basis), dtype=torch.bool)
for iter in range(self.n_iter):
    # Select which (coefficient, basis function) pair to update using the inner product.
    candidate_coefs = residual @ dictionary  # [batch_size, n_basis]
    top_coef_idxs = torch.argmax(torch.abs(candidate_coefs) / dictionary_norms, dim=1)  # [batch_size]
    used_basis_fns[torch.arange(batch_size), top_coef_idxs] = True

    # Update the coefficients.
    used_dictionary = dictionary[..., used_basis_fns].reshape((n_features, batch_size, iter + 1))
    (used_coefficients, _, _, _) = torch.linalg.lstsq(
        used_dictionary.permute((1, 0, 2)),  # [batch_size, n_features, iter + 1]
        residuals.reshape(batch_size, n_features, 1),
    )  # [batch_size, iter + 1, 1]
    coefficients[used_basis_fns] = used_coefficients.reshape(-1)

    # Update the residual.
    residual = data - coefficients @ dictionary.T  # [batch_size, n_features]
```

File: sparsecoding/inference.py [hunk 2]
```diff
@@ -670,3 +670,269 @@ def infer(self, data, dictionary, coeff_0=None):
...
+        device = dictionary.device
+
+        # Define signal sparsity
+        K = np.ceil(self.sparsity*n_basis).astype(int)
+
+        # Initialize coefficients for the whole batch
+        coefficients = torch.zeros(
+            batch_size, n_basis, requires_grad=False, device=device)
+
+        # For each sample in the batch
```
[the remark this answers is not in the document]
Comment (denkle) [reply sparsecoding/inference.py:732]: coefficients = coefficients + update ?

[the remark this answers is not in the document]
Comment (alvinzz) [reply sparsecoding/inference.py:732]: Oh yes, that is probably correct, thank you.

**Comment by alvinzz** [comment]
`autopep8 --max-line-length=120 --in-place --recursive ./` will auto-format the repository according to the `flake8` formatting guidelines.

File: sparsecoding/inference.py [hunk 3]
```diff
@@ -670,3 +670,269 @@ def infer(self, data, dictionary, coeff_0=None):
...
+    Method description can be traced to "Iterative Hard Thresholding for Compressed Sensing" (T. Blumensath & M. E. Davies, 2009)
+    """
+
+    def __init__(self, sparsity, n_iter=10, solver=None, return_all_coefficients=False):
+        '''
+
+        Parameters
+        ----------
+        sparsity : scalar (1,)
+            sparsity of the solution
```
[the remark this answers is not in the document]
Comment (denkle) [reply sparsecoding/inference.py]: Added

**Pull request merged**